rust-runtime/aws-smithy-http-server/src/request/lambda.rs

```diff
@@ -12,41 +12,111 @@ pub use lambda_http::{
     aws_lambda_events::apigw::{ApiGatewayProxyRequestContext, ApiGatewayV2httpRequestContext},
     Context,
 };
+use thiserror::Error;
 
-use super::{extension::MissingExtension, FromParts};
-use crate::Extension;
+use super::{internal_server_error, FromParts};
+use crate::{body::BoxBody, response::IntoResponse};
+
+/// The [`Context`] was not found in the [`http::Request`] extensions.
+///
+/// Use [`LambdaHandler`](crate::routing::LambdaHandler) to ensure it's present.
+#[non_exhaustive]
+#[derive(Debug, Error)]
+#[error("`Context` is not present in the `http::Request` extensions - consider using `LambdaHandler`")]
+pub struct MissingContext;
+
+impl<Protocol> IntoResponse<Protocol> for MissingContext {
+    fn into_response(self) -> http::Response<BoxBody> {
+        internal_server_error()
+    }
+}
 
 impl<P> FromParts<P> for Context {
-    type Rejection = MissingExtension;
+    type Rejection = MissingContext;
 
     fn from_parts(parts: &mut http::request::Parts) -> Result<Self, Self::Rejection> {
-        let Extension(context) = <Extension<Self> as FromParts<P>>::from_parts(parts)?;
-        Ok(context)
+        parts.extensions.remove().ok_or(MissingContext)
+    }
+}
+
+#[derive(Debug, Error)]
+enum MissingGatewayContextTypeV1 {
```

Comment:
We could probably reuse this in place of `MissingGatewayContextTypeV2` if we wanted to - I'm not sure if this would amount to less code however.